components/web_discovery/browser/web_discovery_service.cc

```diff
@@ -0,0 +1,117 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/web_discovery/browser/web_discovery_service.h"
+
+#include <utility>
+
+#include "base/feature_list.h"
+#include "base/functional/bind.h"
+#include "brave/components/constants/pref_names.h"
+#include "brave/components/web_discovery/browser/pref_names.h"
+#include "brave/components/web_discovery/browser/server_config_loader.h"
+#include "brave/components/web_discovery/common/features.h"
+#include "components/prefs/pref_registry_simple.h"
+#include "components/prefs/pref_service.h"
+#include "components/prefs/scoped_user_pref_update.h"
+#include "extensions/buildflags/buildflags.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/service_manager/public/cpp/interface_provider.h"
+
+namespace web_discovery {
+
+WebDiscoveryService::WebDiscoveryService(
+    PrefService* local_state,
+    PrefService* profile_prefs,
+    base::FilePath user_data_dir,
+    scoped_refptr<network::SharedURLLoaderFactory> shared_url_loader_factory)
+    : local_state_(local_state),
+      profile_prefs_(profile_prefs),
+      user_data_dir_(user_data_dir),
+      shared_url_loader_factory_(shared_url_loader_factory) {
+#if BUILDFLAG(ENABLE_EXTENSIONS)
+  if (profile_prefs_->GetBoolean(kWebDiscoveryExtensionEnabled)) {
+    profile_prefs_->ClearPref(kWebDiscoveryExtensionEnabled);
+    profile_prefs_->SetBoolean(kWebDiscoveryNativeEnabled, true);
+  }
+#endif
+
+  pref_change_registrar_.Init(profile_prefs);
+  pref_change_registrar_.Add(
```

Comment:
make sure to implement `KeyedService::Shutdown()` and perform all pref unsubscription along with server config and credential manager reset.